<h1>Содержание<span class="tocSkip"></span></h1>
<div class="toc"><ul class="toc-item"><li><span><a href="#Загрузка-и-изучение-данных" data-toc-modified-id="Загрузка-и-изучение-данных-1"><span class="toc-item-num">1&nbsp;&nbsp;</span>Загрузка и изучение данных</a></span><ul class="toc-item"><li><span><a href="#Общая-информация-о-файле" data-toc-modified-id="Общая-информация-о-файле-1.1"><span class="toc-item-num">1.1&nbsp;&nbsp;</span>Общая информация о файле</a></span></li><li><span><a href="#Предобработка-данных" data-toc-modified-id="Предобработка-данных-1.2"><span class="toc-item-num">1.2&nbsp;&nbsp;</span>Предобработка данных</a></span></li></ul></li><li><span><a href="#Обучение" data-toc-modified-id="Обучение-2"><span class="toc-item-num">2&nbsp;&nbsp;</span>Обучение</a></span><ul class="toc-item"><li><span><a href="#Подготовка-признаков-к-обучению-моделей" data-toc-modified-id="Подготовка-признаков-к-обучению-моделей-2.1"><span class="toc-item-num">2.1&nbsp;&nbsp;</span>Подготовка признаков к обучению моделей</a></span></li><li><span><a href="#CountVectorizer" data-toc-modified-id="CountVectorizer-2.2"><span class="toc-item-num">2.2&nbsp;&nbsp;</span>CountVectorizer</a></span><ul class="toc-item"><li><span><a href="#Линейная-регрессия" data-toc-modified-id="Линейная-регрессия-2.2.1"><span class="toc-item-num">2.2.1&nbsp;&nbsp;</span>Линейная регрессия</a></span></li><li><span><a href="#CatBoostClassifier" data-toc-modified-id="CatBoostClassifier-2.2.2"><span class="toc-item-num">2.2.2&nbsp;&nbsp;</span>CatBoostClassifier</a></span></li><li><span><a href="#LGBMClassifier" data-toc-modified-id="LGBMClassifier-2.2.3"><span class="toc-item-num">2.2.3&nbsp;&nbsp;</span>LGBMClassifier</a></span></li><li><span><a href="#SGDClassifier" data-toc-modified-id="SGDClassifier-2.2.4"><span class="toc-item-num">2.2.4&nbsp;&nbsp;</span>SGDClassifier</a></span></li><li><span><a href="#XGBClassifier" data-toc-modified-id="XGBClassifier-2.2.5"><span class="toc-item-num">2.2.5&nbsp;&nbsp;</span>XGBClassifier</a></span></li></ul></li><li><span><a href="#TF-IDF" data-toc-modified-id="TF-IDF-2.3"><span class="toc-item-num">2.3&nbsp;&nbsp;</span>TF-IDF</a></span><ul class="toc-item"><li><span><a href="#Линейная-регрессия" data-toc-modified-id="Линейная-регрессия-2.3.1"><span class="toc-item-num">2.3.1&nbsp;&nbsp;</span>Линейная регрессия</a></span></li><li><span><a href="#CatBoostClassifier" data-toc-modified-id="CatBoostClassifier-2.3.2"><span class="toc-item-num">2.3.2&nbsp;&nbsp;</span>CatBoostClassifier</a></span></li><li><span><a href="#LGBMClassifier" data-toc-modified-id="LGBMClassifier-2.3.3"><span class="toc-item-num">2.3.3&nbsp;&nbsp;</span>LGBMClassifier</a></span></li><li><span><a href="#SGDClassifier" data-toc-modified-id="SGDClassifier-2.3.4"><span class="toc-item-num">2.3.4&nbsp;&nbsp;</span>SGDClassifier</a></span></li><li><span><a href="#XGBClassifier" data-toc-modified-id="XGBClassifier-2.3.5"><span class="toc-item-num">2.3.5&nbsp;&nbsp;</span>XGBClassifier</a></span></li></ul></li></ul></li><li><span><a href="#Проверка-моделей-на-тестовой-выборке" data-toc-modified-id="Проверка-моделей-на-тестовой-выборке-3"><span class="toc-item-num">3&nbsp;&nbsp;</span>Проверка моделей на тестовой выборке</a></span><ul class="toc-item"><li><span><a href="#Тестированиние-моделей-построенных-с-помощью-&quot;мешка-слов&quot;" data-toc-modified-id="Тестированиние-моделей-построенных-с-помощью-&quot;мешка-слов&quot;-3.1"><span class="toc-item-num">3.1&nbsp;&nbsp;</span>Тестированиние моделей построенных с помощью "мешка слов"</a></span></li><li><span><a href="#Тестирование-моделей-созданных-с-помощью-TF-IDF" data-toc-modified-id="Тестирование-моделей-созданных-с-помощью-TF-IDF-3.2"><span class="toc-item-num">3.2&nbsp;&nbsp;</span>Тестирование моделей созданных с помощью TF-IDF</a></span></li></ul></li><li><span><a href="#Общий-вывод" data-toc-modified-id="Общий-вывод-4"><span class="toc-item-num">4&nbsp;&nbsp;</span>Общий вывод</a></span></li></ul></div>

# Проект для «Викишоп»

Интернет-магазин «Викишоп» запускает новый сервис. Теперь пользователи могут редактировать и дополнять описания товаров, как в вики-сообществах. То есть клиенты предлагают свои правки и комментируют изменения других. Магазину нужен инструмент, который будет искать токсичные комментарии и отправлять их на модерацию. 

Обучите модель классифицировать комментарии на позитивные и негативные. В вашем распоряжении набор данных с разметкой о токсичности правок.

Постройте модель со значением метрики качества *F1* не меньше 0.75. 

**Инструкция по выполнению проекта**

1. Загрузите и подготовьте данные.
2. Обучите разные модели. 
3. Сделайте выводы.

Для выполнения проекта применять *BERT* необязательно, но вы можете попробовать.

**Описание данных**

Данные находятся в файле `toxic_comments.csv`. Столбец *text* в нём содержит текст комментария, а *toxic* — целевой признак.

## Загрузка и изучение данных

### Общая информация о файле

Загрузим все необходимые библиотеки.

In [1]:
!pip install nltk

In [2]:
import pandas as pd
import numpy as np
import nltk

from nltk.stem import WordNetLemmatizer
import re
from nltk.corpus import stopwords
from nltk.corpus import wordnet
from nltk.stem import WordNetLemmatizer
from nltk.tokenize import word_tokenize 
from nltk import pos_tag

from sklearn.feature_extraction.text import CountVectorizer
from sklearn.feature_extraction.text import TfidfVectorizer

from sklearn.model_selection import train_test_split
from sklearn.metrics import f1_score
from sklearn.linear_model import LogisticRegression
from catboost import CatBoostClassifier
from lightgbm import LGBMClassifier
import lightgbm as lgb
from sklearn.linear_model import SGDClassifier
from xgboost import XGBClassifier


import warnings
warnings.filterwarnings('ignore')

Загрузим файл и посмотрим первые 5 строк.

In [3]:
try:
    df = pd.read_csv('/datasets/toxic_comments.csv') # загрузка данных
except:
    df = pd.read_csv('C:\\Users\\koten\\Downloads\\обратотка текстов\\toxic_comments.csv') # загрузка данных

In [4]:
print(df.head()) # просмотр первых 5 строк

                                                text  toxic
0  Explanation\nWhy the edits made under my usern...      0
1  D'aww! He matches this background colour I'm s...      0
2  Hey man, I'm really not trying to edit war. It...      0
3  "\nMore\nI can't make any real suggestions on ...      0
4  You, sir, are my hero. Any chance you remember...      0


Посмотрим информацию о файле.

In [5]:
df.info() # получение информации о файле

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 159571 entries, 0 to 159570
Data columns (total 2 columns):
 #   Column  Non-Null Count   Dtype 
---  ------  --------------   ----- 
 0   text    159571 non-null  object
 1   toxic   159571 non-null  int64 
dtypes: int64(1), object(1)
memory usage: 2.4+ MB


В таблице всего два столбца: столбец text содержит текст комментария, а toxic — целевой признак. Тип данных у текста object, а у целевого признака int64. Далее посмотрим сколько у нас комментариев, помеченных как токсичные и сколько обычных.

In [6]:
df['toxic'].value_counts() # подсчет количества занчений по классам

0    143346
1     16225
Name: toxic, dtype: int64

Мы видим сильный дисбалланс классов. Это стоит учесть при обучении моделей.

### Предобработка данных

Прежде чем приступить с обучению, нужно данные очистить от неинформативных символов и слов, а также привести слова к их начальной форме (провести лемматизацию).

В нашем распоряжении текст на английском языке, поэтому для лемматизации будем использовать WordnetLemmatizer. А для улучшения качества лемматизируемого текста мы воспользуемся POS-тегирование, которое отнесет каждое слово к определенной части речи. Для того, чтобы найти правильный POS-тег для каждого слова, сопоставим его с правильным входным символом, который принимает WordnetLemmatizer, и передадим его в качестве второго аргумента в lemmatize().
В nltk для этого есть метод nltk.pos_tag(). Он принимает список слов, а возвращает кортеж с тегом POS. Напишем функцию для POS-тегирования.

In [7]:
def get_wordnet_pos(word): # функция для создания POS-тегов 
    tag = nltk.pos_tag([word])[0][1][0].upper()
    tag_dict = {"J": wordnet.ADJ,
                "N": wordnet.NOUN,
                "V": wordnet.VERB,
                "R": wordnet.ADV}
    return tag_dict.get(tag, wordnet.NOUN)

Далее мы сделаем следующее:
Напишем функцию, которая:
- приведет текст к нижнему регистру
- удалит лишние символы
- избавится от лишних пробелов

В следующем этапе создадим список для хранения преобразованных данных, загрузим словарь стоп-слов и инициализируем лемматизатор.


С помощью цикла пройдемся по столбцу с исходным текстом:
- очистим текст
- токенизируем
- удалим стоп-слова
- лемматизируем с учетом POS-тегов

И в заключении создадим столбец с обработанным текстом.

In [8]:
%%time
nltk.download('averaged_perceptron_tagger')
nltk.download('wordnet')
def clean_text(text):
   
    text = text.lower()  # приводим текст к нижнему регистру
    text = re.sub(r'[^a-zA-Z]', ' ', text)  # удаляем лишние символы    
    text = " ".join(text.split()) # удаляем лишние пробелы
    
    return text # возвращаем очищенные данные
 

processed_text = [] # создаем список для хранения преобразованных данных

stop_words = stopwords.words('english') # загружаем стоп-слова для английского языка

lemmatizer = WordNetLemmatizer() # инициализируем лемматайзер
 

for text in df['text']: # для каждого сообщения text из столбца data['text']
    # cleaning 
    text = clean_text(text)  # очистка текста    
    text = word_tokenize(text) # токенизирование     
    text = [word for word in text if word not in stop_words] # удаление стоп-слов  
    text = ' '.join([lemmatizer.lemmatize(w,get_wordnet_pos(w)) for w in text]) # лемматизация
     
    
    processed_text.append(text) # добавляем преобразованный текст в список processed_text
 

df['lemm_text'] = processed_text # Сохраняем результат преобразования в новой колонке 'lemm_text'

[nltk_data] Downloading package averaged_perceptron_tagger to
[nltk_data]     C:\Users\koten\AppData\Roaming\nltk_data...
[nltk_data]   Package averaged_perceptron_tagger is already up-to-
[nltk_data]       date!
[nltk_data] Downloading package wordnet to
[nltk_data]     C:\Users\koten\AppData\Roaming\nltk_data...
[nltk_data]   Package wordnet is already up-to-date!


Wall time: 27min 40s


На примере одного соощения проверим результат нашего преобразования.

In [9]:
print("Исходный текст:",df['text'][0])
print("Лемматизированный текст:", df['lemm_text'][0])

Исходный текст: Explanation
Why the edits made under my username Hardcore Metallica Fan were reverted? They weren't vandalisms, just closure on some GAs after I voted at New York Dolls FAC. And please don't remove the template from the talk page since I'm retired now.89.205.38.27
Лемматизированный текст: explanation edits make username hardcore metallica fan revert vandalism closure gas vote new york doll fac please remove template talk page since retire


Преобразовани проведены достаточно успешно: весь текст написан в нижнем регистре, между словани один пробел, нет знаков препинания, а также удалены неинформативные для модели слова.

**Вывод**

В первом разделе нашего проекта мы посмотрели исходные данные. В нашем распоряжении таблица из двух столбцов: в первом текст комментария, а во втором пометка о том токсичный комментарий или нет. Также мы обнаружили дисбалланс классов, который следует учесть при обучении моделей.

Далее мы провели предобработку текста:
- очистили текст от ненужных символов
- токенизировали
- удалили стоп-слова (слова, которые не важны для обучения моделей)
- лемматизировали текст с учетом POS-тегов (привели слова к начальной форме).

## Обучение

### Подготовка признаков к обучению моделей

Разделим выборку на признаки и целевой признаки.

In [10]:
features = df['lemm_text'] # признаки
target = df['toxic'] # целевой признак

Разделим данные на обучающую, валидационную и тестовую выборку. Делить будем в соотношении 60/20/20.

In [11]:
features_train, features_valid_test, target_train, target_valid_test = train_test_split(features, target, test_size=0.4, random_state=12345)
features_valid, features_test, target_valid, target_test = train_test_split(features_valid_test, target_valid_test, test_size=0.5, random_state=12345)


features_train.shape, features_valid.shape, features_test.shape, target_train.shape, target_valid.shape, target_test.shape

((95742,), (31914,), (31915,), (95742,), (31914,), (31915,))

Проверим сохраняется ли дисбалланс классов при разделении на подвыборки.

In [12]:
print('Распределение классов в обучающей выборке:')
print(target_train.value_counts()[0] / target_train.value_counts().sum())
print(target_train.value_counts()[1] / target_train.value_counts().sum())
print()
print('Распределение классов в обучающей выборке:')
print(target_valid.value_counts()[0] / target_valid.value_counts().sum())
print(target_valid.value_counts()[1] / target_valid.value_counts().sum())
print()
print('Распределение классов в тестовой выборке:')
print(target_test.value_counts()[0] / target_test.value_counts().sum())
print(target_test.value_counts()[1] / target_test.value_counts().sum())

Распределение классов в обучающей выборке:
0.8981324810428025
0.10186751895719746

Распределение классов в обучающей выборке:
0.8977878047251989
0.10221219527480102

Распределение классов в тестовой выборке:
0.8994203352655491
0.10057966473445089


Создадим корпуса слов для каждой выборки.

In [13]:
train_corpus = features_train.values.astype('U') #создание корпуса слов на обучающей выборке
valid_corpus = features_valid.values.astype('U') #создание корпуса слов на валидационной выборке
test_corpus = features_test.values.astype('U') #создание корпуса слов на тестовой выборке

Теперь наши данные готовы. Далее мы применем два подхода: "мешок слов" и TF-IDT преобразования слов в числа. Для каждого подхода обучим несколько моделей. Учитывая дисбалланс классов при обучении моделей будем использовать 'class_weight'

### CountVectorizer

Переведём тексты в понятный для машины формат — векторный. Преобразовать слова в векторы поможет модель «мешок слов». Для того чтобы преобразовать корпус текстов в мешок слов, обратимся к классу CountVectorizer(). Применим метод fit(), только к обучающей выборке, иначе тестирование будет нечестным: в модели будут учтены частоты слов из тестовой и валидационной выборок.

In [14]:
count_vect = CountVectorizer(dtype=np.float32) #создаем счетчик
bow_train = count_vect.fit_transform(train_corpus) #счетчик выделяет из корпуса уникальные слова и подсчитывает количество их вхождений в каждом тексте корпуса
bow_valid = count_vect.transform(valid_corpus)
bow_test = count_vect.transform(test_corpus)

Далее обучим модели линейной регрессии, CatBoostClassifier, LGBMClassifier, SGDRClassifier и XGBClassifier. Метрикой качества будет F1-мера.                                         '

#### Линейная регрессия


In [15]:
%%time

model_lr_bow = LogisticRegression(random_state=12345, solver='liblinear', class_weight='balanced')
model_lr_bow.fit(bow_train, target_train)

predicted_valid = model_lr_bow.predict(bow_valid)
f1_lr_bow = f1_score(target_valid, predicted_valid) 
print('Логистическая регрессия')
print('F1: {:.3f}'.format(f1_lr_bow))

Логистическая регрессия
F1: 0.744
Wall time: 6.69 s


F1-мера равна 0.744, а время обучения в пределах 10 секунд, что достаточно быстро, но метрика качества немного не дотягивает до порога в 0.75. Посмотрим как поведут сябя другие модели.

#### CatBoostClassifier

In [16]:
%%time

model_cat_bow = CatBoostClassifier(random_state=12345, auto_class_weights = 'Balanced', custom_metric='F1')
model_cat_bow.fit(bow_train, target_train, verbose=50)

predicted_valid = model_cat_bow.predict(bow_valid)
f1_cat_bow = f1_score(target_valid, predicted_valid)
print('CatBoost:')
print('F1: {:.3f}'.format(f1_cat_bow))

Learning rate set to 0.072254
0:	learn: 0.6556605	total: 557ms	remaining: 9m 16s
50:	learn: 0.4268949	total: 21s	remaining: 6m 30s
100:	learn: 0.3826242	total: 39.8s	remaining: 5m 54s
150:	learn: 0.3579584	total: 58.6s	remaining: 5m 29s
200:	learn: 0.3310534	total: 1m 17s	remaining: 5m 7s
250:	learn: 0.3119354	total: 1m 36s	remaining: 4m 47s
300:	learn: 0.2967121	total: 1m 54s	remaining: 4m 26s
350:	learn: 0.2846078	total: 2m 13s	remaining: 4m 7s
400:	learn: 0.2744319	total: 2m 32s	remaining: 3m 47s
450:	learn: 0.2661813	total: 2m 50s	remaining: 3m 27s
500:	learn: 0.2589913	total: 3m 9s	remaining: 3m 8s
550:	learn: 0.2522776	total: 3m 27s	remaining: 2m 49s
600:	learn: 0.2463989	total: 3m 46s	remaining: 2m 30s
650:	learn: 0.2406820	total: 4m 4s	remaining: 2m 11s
700:	learn: 0.2360708	total: 4m 23s	remaining: 1m 52s
750:	learn: 0.2314416	total: 4m 41s	remaining: 1m 33s
800:	learn: 0.2273370	total: 5m	remaining: 1m 14s
850:	learn: 0.2228581	total: 5m 20s	remaining: 56s
900:	learn: 0.21927

Здесь F1 уже 0,753, а время обучения увеличилось до 6 минут.

#### LGBMClassifier

In [17]:
%%time

       
model_lgbm_bow = LGBMClassifier(random_state=12345, class_weight='balanced')           
model_lgbm_bow.fit(bow_train, target_train)

predicted_valid = model_lgbm_bow.predict(bow_valid)

f1_lgbm_bow = f1_score(target_valid, predicted_valid)
print('LGBMClassifier:')
print('F1: {:.3f}'.format(f1_lgbm_bow))

LGBMClassifier:
F1: 0.740
Wall time: 4.31 s


Время обучения немного меньше, чем у CatBoost, но и F1 всего 0.74

#### SGDClassifier

In [18]:
%%time

model_sdg_bow = SGDClassifier(random_state=12345, class_weight='balanced')
model_sdg_bow.fit(bow_train, target_train)

predicted_valid = model_sdg_bow.predict(bow_valid)
f1_sdg_bow = f1_score(target_valid, predicted_valid)
print('LGBMClassifier:')
print('F1: {:.3f}'.format(f1_sdg_bow))


LGBMClassifier:
F1: 0.751
Wall time: 1.13 s


Хорошее сочетание срорости обучения (около минуты) и размера F1 - 0.751

#### XGBClassifier

In [19]:
%%time

model_xgb_bow = XGBClassifier(random_state=12345, scale_pos_weight = 10)
model_xgb_bow.fit(bow_train, target_train)
          
predicted_valid = model_xgb_bow.predict(bow_valid)
f1_xgb_bow = f1_score(target_valid, predicted_valid)
print('XGBClassifier:')
print('F1: {:.3f}'.format(f1_xgb_bow))

XGBClassifier:
F1: 0.734
Wall time: 4.66 s


Хуже всего справилась с задачей модель XGB: F1 всего 0.734, при этом на обучение ушло почти пять минут.

### TF-IDF

Далее посмотрим другой подход к преобразованию текстов в вектора. Мешок слов учитывает частоту употребления слов, а мы посчитаем, как часто уникальное слово встречается во всём корпусе и в отдельном его тексте и определим "важность" слова. Оценка важности слова определяется величиной TF-IDF.

Обучать мы будем те же модели, что и с "мешком слов".

In [20]:
count_tf_idf = TfidfVectorizer() # создаем счетчик

tf_idf_train = count_tf_idf.fit_transform(train_corpus)  # расчет TF-IDF для корпуса текстов
tf_idf_valid = count_tf_idf.transform(valid_corpus)
tf_idf_test = count_tf_idf.transform(test_corpus)

#### Линейная регрессия

In [21]:
%%time

model_lr_tf_idf = LogisticRegression(random_state=12345, solver='liblinear', class_weight='balanced')
model_lr_tf_idf.fit(tf_idf_train, target_train)

predicted_valid = model_lr_tf_idf.predict(tf_idf_valid)

f1_lr_tf_idf = f1_score(target_valid, predicted_valid) 
print('Логистическая регрессия')
print('F1: {:.3f}'.format(f1_lr_tf_idf))


Логистическая регрессия
F1: 0.753
Wall time: 953 ms


F1 равна 0,753 на валидационной выборке. Время обучения очень маленькое.

#### CatBoostClassifier

In [22]:
%%time

model_cat_tf_idf = CatBoostClassifier(random_state=12345, auto_class_weights = 'Balanced', )
model_cat_tf_idf.fit(tf_idf_train, target_train, verbose=50)

predicted_valid = model_cat_tf_idf.predict(tf_idf_valid)
                     
f1_cat_tf_idf = f1_score(target_valid, predicted_valid)
print('CatBoost:')
print('F1: {:.3f}'.format(f1_cat_tf_idf))

Learning rate set to 0.072254
0:	learn: 0.6554324	total: 1.06s	remaining: 17m 37s
50:	learn: 0.4252444	total: 34.1s	remaining: 10m 34s
100:	learn: 0.3778538	total: 1m 6s	remaining: 9m 48s
150:	learn: 0.3491063	total: 1m 38s	remaining: 9m 11s
200:	learn: 0.3236512	total: 2m 9s	remaining: 8m 36s
250:	learn: 0.3029473	total: 2m 41s	remaining: 8m 2s
300:	learn: 0.2874020	total: 3m 13s	remaining: 7m 28s
350:	learn: 0.2745307	total: 3m 44s	remaining: 6m 55s
400:	learn: 0.2628368	total: 4m 16s	remaining: 6m 22s
450:	learn: 0.2531798	total: 4m 47s	remaining: 5m 50s
500:	learn: 0.2447450	total: 5m 19s	remaining: 5m 17s
550:	learn: 0.2369723	total: 5m 50s	remaining: 4m 45s
600:	learn: 0.2299987	total: 6m 21s	remaining: 4m 13s
650:	learn: 0.2235184	total: 6m 53s	remaining: 3m 41s
700:	learn: 0.2178135	total: 7m 24s	remaining: 3m 9s
750:	learn: 0.2121319	total: 7m 56s	remaining: 2m 37s
800:	learn: 0.2071866	total: 8m 27s	remaining: 2m 6s
850:	learn: 0.2027119	total: 8m 58s	remaining: 1m 34s
900:	l

F1 равна 0.76, пока что это лучший показатель, но время обучения почти 11 минут.

#### LGBMClassifier

In [23]:
%%time
      
model_lgbm_tf_idf = LGBMClassifier(random_state=12345, class_weight='balanced')           
model_lgbm_tf_idf.fit(tf_idf_train, target_train)

predicted_valid = model_lgbm_tf_idf.predict(tf_idf_valid)

f1_lgbm_tf_idf = f1_score(target_valid, predicted_valid)
print('LGBMClassifier:')
print('F1: {:.3f}'.format(f1_lgbm_tf_idf))

LGBMClassifier:
F1: 0.738
Wall time: 17.8 s


LGBMClassifier в этом случае, плохо справился с поставленной задачей. При очень быстром обучении (около 20 секунд), F1 всего 0.738

#### SGDClassifier

In [24]:
%%time

model_sdg_tf_idf = SGDClassifier(random_state=12345, class_weight='balanced')
model_sdg_tf_idf.fit(tf_idf_train, target_train)

predicted_valid = model_sdg_tf_idf.predict(tf_idf_valid)

f1_sdg_tf_idf = f1_score(target_valid, predicted_valid)
print('LGBMClassifier:')
print('F1: {:.3f}'.format(f1_sdg_tf_idf))

LGBMClassifier:
F1: 0.740
Wall time: 403 ms


F1 равна 0.740, что ниже установленного минимального показателя качества.

#### XGBClassifier

In [25]:
%%time


model_xgb_tf_idf = XGBClassifier(random_state=12345, scale_pos_weight = 10)
model_xgb_tf_idf.fit(tf_idf_train, target_train)
          
predicted_valid = model_xgb_tf_idf.predict(tf_idf_valid)
f1_xgb_tf_idf = f1_score(target_valid, predicted_valid)
print('XGBClassifier:')
print('F1: {:.3f}'.format(f1_xgb_tf_idf))

XGBClassifier:
F1: 0.723
Wall time: 16.4 s


И опять модель XGB показала худший результат: F1 всего 0,723

Сведем результаты в одну таблицу.

In [26]:
score_data = {'F1 мешок слов': [f1_lr_bow, f1_cat_bow, f1_lgbm_bow, f1_sdg_bow, f1_xgb_bow ],
                       'F1 TF-IDF': [f1_lr_tf_idf, f1_cat_tf_idf, f1_lgbm_tf_idf, f1_sdg_tf_idf, f1_xgb_tf_idf]}
score_df = pd.DataFrame(score_data, index = ['Линейная регрессия',
                                            'CatBoostClassifier',
                                             'LGBMClassifier',
                                            'SGDRClassifier','XGBClassifier'])
score_df['F1 мешок слов'] = round(score_df['F1 мешок слов'],3)
score_df['F1 TF-IDF'] = round(score_df['F1 TF-IDF'],3)
print('Результаты F1  моделей')
display(score_df)

Результаты F1  моделей


F1 мешок слов  F1 TF-IDF
Линейная регрессия          0.744      0.753
CatBoostClassifier          0.753      0.760
LGBMClassifier              0.740      0.738
SGDRClassifier              0.751      0.740
XGBClassifier               0.734      0.723

При векторизации текста с помощью "мешка слов" лучший результат на валидационной выборке показала модель CatBoostClassifier (F1 = 0.753), на втором месте SGDRClassifier (F1 = 0,751). При применении TF-IDF лучший результат на валидационной выборке опять у CatBoostClassifier (F1 = 0,76), а на втором месте линейная регрессия (F1 = 0.753).

**Вывод**

В этом разделе мы посмотрели два варианта преобразования текства в вектор: "мешок слов" и TF-IDF. На преобразованных признаках обучили модели линейной регрессии, CatBoostClassifier, LGBMClassifier, SGDRClassifier и XGBClassifier. Проверили качество на валидационной выборке. В этом проекте метрикой качества является F1-мера.

Лучший результат в обоих случаях показала модель CatBoostClassifier, а худший XGBClassifier. Далее проверим качество моделей на тестовой выборке.

## Проверка моделей на тестовой выборке

### Тестированиние моделей построенных с помощью "мешка слов"

In [27]:
predicted_test = model_lr_bow.predict(bow_test) # линейная регрессия ("мешок слов")
f1_lr_bow_test = f1_score(target_test, predicted_test) 
print('Логистическая регрессия')
print('F1: {:.3f}'.format(f1_lr_bow_test))

Логистическая регрессия
F1: 0.740


In [28]:
predicted_test = model_cat_bow.predict(bow_test) # CatBoostClassifier ("мешок слов")
f1_cat_bow_test = f1_score(target_test, predicted_test)
print('CatBoostClassifier:')
print('F1: {:.3f}'.format(f1_cat_bow_test))

CatBoostClassifier:
F1: 0.747


In [29]:
predicted_test = model_lgbm_bow.predict(bow_test) # LGBMClassifier ("мешок слов")

f1_lgbm_bow_test = f1_score(target_test, predicted_test)
print('LGBMClassifier:')
print('F1: {:.3f}'.format(f1_lgbm_bow_test))

LGBMClassifier:
F1: 0.738


In [30]:
predicted_test = model_sdg_bow.predict(bow_test) # SGDRClassifier ("мешок слов")
f1_sdg_bow_test = f1_score(target_test, predicted_test)
print('SGDRClassifier:')
print('F1: {:.3f}'.format(f1_sdg_bow_test))

SGDRClassifier:
F1: 0.751


In [31]:
predicted_test = model_xgb_bow.predict(bow_test) # XGBClassifier ("мешок слов")
f1_xgb_bow_test = f1_score(target_test, predicted_test)
print('XGBClassifier:')
print('F1: {:.3f}'.format(f1_xgb_bow_test))


XGBClassifier:
F1: 0.733


### Тестирование моделей созданных с помощью TF-IDF

In [32]:
predicted_test = model_lr_tf_idf.predict(tf_idf_test) # линейная регрессия ("TF-IDF")

f1_lr_tf_idf_test = f1_score(target_test, predicted_test) 
print('Логистическая регрессия')
print('F1: {:.3f}'.format(f1_lr_tf_idf_test))


Логистическая регрессия
F1: 0.752


In [33]:
predicted_test = model_cat_tf_idf.predict(tf_idf_test) # CatBoost ("TF-IDF")
                     
f1_cat_tf_idf_test = f1_score(target_test, predicted_test)
print('CatBoost:')
print('F1: {:.3f}'.format(f1_cat_tf_idf_test))

CatBoost:
F1: 0.757


In [34]:
predicted_test = model_lgbm_tf_idf.predict(tf_idf_test) # LGBMClassifier ("TF-IDF")

f1_lgbm_tf_idf_test = f1_score(target_test, predicted_test)
print('LGBMClassifier:')
print('F1: {:.3f}'.format(f1_lgbm_tf_idf_test))

LGBMClassifier:
F1: 0.735


In [35]:
predicted_test = model_sdg_tf_idf.predict(tf_idf_test) # SGDRClassifier ("TF-IDF")

f1_sdg_tf_idf_test = f1_score(target_test, predicted_test)
print('SGDRClassifier:')
print('F1: {:.3f}'.format(f1_sdg_tf_idf_test))

SGDRClassifier:
F1: 0.736


In [36]:
predicted_test = model_xgb_tf_idf.predict(tf_idf_test) # XGBClassifier ("TF-IDF")
f1_xgb_tf_idf_test = f1_score(target_test, predicted_test)
print('XGBClassifier:')
print('F1: {:.3f}'.format(f1_xgb_tf_idf_test))

XGBClassifier:
F1: 0.728


Сведем полученные данные в одну таблицу.

In [37]:
score_data = {'F1 мешок слов': [f1_lr_bow, f1_cat_bow, f1_lgbm_bow, f1_sdg_bow, f1_xgb_bow ],
              'F1 мешок слов test': [f1_lr_bow_test, f1_cat_bow_test, f1_lgbm_bow_test, f1_sdg_bow_test, f1_xgb_bow_test],
             'F1 TF-IDF': [f1_lr_tf_idf, f1_cat_tf_idf, f1_lgbm_tf_idf, f1_sdg_tf_idf, f1_xgb_tf_idf],
             'F1 TF-IDF test': [f1_lr_tf_idf_test, f1_cat_tf_idf_test, f1_lgbm_tf_idf_test, f1_sdg_tf_idf_test, f1_xgb_tf_idf_test]
             }
score_df = pd.DataFrame(score_data, index = ['Линейная регрессия',
                                            'CatBoostClassifier',
                                             'LGBMClassifier',
                                            'SGDRClassifier','XGBClassifier'])
score_df['F1 мешок слов'] = round(score_df['F1 мешок слов'],3)
score_df['F1 мешок слов test'] = round(score_df['F1 мешок слов test'],3)
score_df['F1 TF-IDF'] = round(score_df['F1 TF-IDF'],3)
score_df['F1 TF-IDF test'] = round(score_df['F1 TF-IDF test'],3)
print('Результаты F1  моделей')
display(score_df)

Результаты F1  моделей


F1 мешок слов  F1 мешок слов test  F1 TF-IDF  \
Линейная регрессия          0.744               0.740      0.753   
CatBoostClassifier          0.753               0.747      0.760   
LGBMClassifier              0.740               0.738      0.738   
SGDRClassifier              0.751               0.751      0.740   
XGBClassifier               0.734               0.733      0.723   

                    F1 TF-IDF test  
Линейная регрессия           0.752  
CatBoostClassifier           0.757  
LGBMClassifier               0.735  
SGDRClassifier               0.736  
XGBClassifier                0.728

В сводной таблице видно, что лучший показатель F1 на тестовой выборке при применении "мешка слов" показывает модель SGDRClassifier (F1 0.751). А вот CatBoostClassifier, который на валидационной выборке показывал лучший результат, немного не дотянул до 0.75 (F1 = 0.747). 

Лучший показатель F1 на тестовой выборке при применении TF-IDF показывает модель CatBoostClassifier (F1 =0,757), на втором месте линейная регрессия (F1=0,752).

## Общий вывод

В данном проекте мы обучали модели, которые могли бы искать токсичные комментрии. в нашем распоряжении тексты с комментариями и пометкой, относится ли конкретный текст комментрия к токсичному или нет. 

В первом разделе мы посмотрели общую информацию о файле. Мы обнаружили дисбалланс классов, который учли при обучении моделей. Также сразу столо понятно, что на исходном тексте модель не сможет учиться. Поэтому мы провели предобработку, создав дополнительный столбец, в который поместили текст очищенный от стоп-слов и знаков препинания. Слова привели к начальной форме и  сделали, так чтобы между словами был только один пробел. 

Во втором разделе проекта мы очищенный и лемматизированнй текст перевели в векторы, поскольку модели не понимают слова, им нужны числа. Здесь мы применили два подхода: "мешок слов" и TF-IDF. Для каждого метода обучили модели линейной регрессии, CatBoostClassifier, LGBMClassifier, SGDRClassifier и XGBClassifier. Проверили качество на валидационной выборке. В этом проекте метрикой качества является F1-мера. 

Результат получился следующий: при векторизации текста с помощью "мешка слов" лучший результат на валидационной выборке показала модель CatBoostClassifier (F1 = 0.753), на втором месте SGDRClassifier (F1 = 0,751). При применении TF-IDF лучший результат на валидационной выборке опять у CatBoostClassifier (F1 = 0,76), а на втором месте линейная регрессия (F1 = 0.753). В обоих случаях, худший результат показала модель XGBClassifier.


В третьем разделе мы проверили качество моделей на тестовой выборке. Результат получили такой: лучший показатель F1 на тестовой выборке при применении "мешка слов" показывает модель SGDRClassifier (F1 0.751). 
Лучший показатель F1 на тестовой выборке при применении TF-IDF показывает модель CatBoostClassifier (F1 =0,757), на втором месте линейная регрессия (F1=0,752).
